Put the correct type in StateBag to match Azure go SDK

braunsonm opened the pull request:
The Azure Go SDK expects the tags format to use a pointer to the string for the map value.
The configuration from the builder is not a pointer so when storing in the state bag for
reference in later execution we convert it when creating the StateBag.

Fixes #10012
Fixes #10013

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/packer?pullRequest=10014)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/packer?pullRequest=10014)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/packer?pullRequest=10014).</sub>

braunsonm commented:
This was likely introduced from an update in the Azure Go SDK recently.

nywilken commented on builder/azure/arm/builder.go:
```diff
@@ -390,7 +390,12 @@ func (b *Builder) getBlobAccount(ctx context.Context, client *AzureClient, resou
 	stateBag.Put(constants.AuthorizedKey, b.config.sshAuthorizedKey)
 
-	stateBag.Put(constants.ArmTags, b.config.AzureTags)
+	tagsMap := make(map[string]*string)
+	for key, value := range b.config.AzureTags {
+		tagsMap[key] = &value
+	}
+	// Azure SDK expects string pointers for the values
+	stateBag.Put(constants.ArmTags, tagsMap)
```
PR #9985 introduced a helper function `azcommon.MapToAzureTags` that converts tags from a simple map[string]string to a map[string]*string. 

This was not caught by the tests for step_create_resource because test for that step create the expected type. We need to add some test for this case and then update this line to look like 
```suggestion
   stateBag.Put(constants.ArmTags, azcommon.MapToAzureTags(b.config.AzureTags))
```

With that said we should probably add a type assertion check on [here](https://github.com/hashicorp/packer/blob/bdb1c47532655ceac13a14d06bc7e228b33db983/builder/azure/arm/step_create_resource_group.go#L64) to capture any failed conversions which will also prevent the Packer crash in #10012 . 

Something like

```go
tags, ok := state.Get(constants.ArmTags).(map[string]*string)
if !ok {                                        
    err := fmt.Errorf("failed to extract tags from state bag")                                                                                                                                                                                                                                      
    state.Put(constants.Error, err)                                                                                                                                                                                                                                                                    
    s.error(err)                                                                                                                                                                                                                                                                                  
    return multistep.ActionHalt                                                                                                                                                                                                                                                                        
  } 
```

nywilken reviewed (changes requested):
Hello again thanks for opening up this PR. It is on the right track. There is a helper that should do what your code is doing. To keep the logic consistent the change should move to using the helper. 

Some additional test cases need to be added, possibly by just updating the existing to use the config and not a raw map[string]*string to prevent this in the future. We, the Packer maintainers, would like to get this fix rolled out as soon as possible so I am looking into the tests now. But if you could make the change to use the helper and test again on your end that would be great.

braunsonm commented:
Hey @nywilken addressed your comments as you stated you're taking a look at the tests.

I would also like to roll this out as soon as possible. Why is there no integration tests in azure that would have caught this?

braunsonm commented:
Not sure if you were looking at adding a test change to the builder but I added a case on the resource group creation.

nywilken commented:
> Not sure if you were looking at adding a test change to the builder but I added a case on the resource group creation. This change is working on my end with a test packer config.

Thanks for the quick turnaround @braunsonm adding a test to the builder is what I was looking into as that is where we make the initial conversion. The test you have in place looks great. Interesting enough it looks like result is always multistep.ActionHalt for those test. But fails on master which is what we want. 

I wrote a quick test for the builder that requires two changes would you like for me to add as a comment or can I add as a commit to your branch? 

The acceptance tests that accounts for tags does not seem to be a thing so we should add that. As to why there are no tests already, slight oversight. We've have been working on improving acceptance tests across the builders. So this is a step in the right direction.

braunsonm commented:
Sounds good, go ahead and commit to this branch @nywilken

nywilken reviewed (approved):
> Sounds good, go ahead and commit to this branch @nywilken

@braunsonm thanks for letting me push to your branch. I added two commits which should capture the additional tests. Going to get a second pair of eyes on the commits but this should be good to go. We are anticipating a 1.6.4 this week to address this regression along with #10000 

My only nit is the go.mod files I think those changes are from building the project. But if it needs to be dropped I'll square that away at merge time.

nywilken commented:
@azr could you take a quick look over the changes. The changes have been tested with both JSON and HCL configs.

nywilken commented:
@braunsonm if I didn't say it already. Thanks for working with me on getting this patched up. Great stuff!

braunsonm commented:
First time using Packer today to find it odd that it wasn't working 😂 Happy to help!

azr commented on builder/azure/arm/builder.go:
```diff
@@ -390,7 +390,7 @@ func (b *Builder) getBlobAccount(ctx context.Context, client *AzureClient, resou
 	stateBag.Put(constants.AuthorizedKey, b.config.sshAuthorizedKey)
 
-	stateBag.Put(constants.ArmTags, b.config.AzureTags)
+	stateBag.Put(constants.ArmTags, packerAzureCommon.MapToAzureTags(b.config.AzureTags))
```
ah bummer; I had missed that one ! LGTM ! Good fix !
